### PR TITLE
Convert private raw pointers to unique_ptr

### PR DESCRIPTION
**Pull request by neheb**

### CODE REVIEW

**Comment by neheb** [comment]
as it stands, List does not support move semantics. I tried before to add but was unsuccessful.

**Comment by sbooth** [comment]
@scotchi TagLib 2.0 could be a good opportunity to switch to the standard containers. I haven't investigated the feasibility of switching but I do think it's worth considering as long as the needed features are present.

**Comment by scotchi** [comment]
I'm not conceptually opposed to such, but it'd be a massive change and break source compatibility in a pretty huge way.

**Comment by neheb** [comment]
Do note I'm not touching headers. This is all in cpp. No API is being changed.

**Comment by scotchi** [comment]
I was responding to @sbooth.  I realized that this change is just in the .cpp file, but it's still not something that's done anywhere else in TagLib, and I'm not thrilled about introducing the inconsistency.

**Comment by ufleisch** [comment]
I would also like to keep using the TagLib containers. I prefer implicitly shared container classes as provided by Qt and TagLib over the STL containers because they are easier to use.

**Comment by scotchi** [comment]
Note that #1122 already got merged, and should be considered along with this.

**Comment by neheb** [comment]
Rebased.

**Comment by CyberSinh** [comment]
Would this PR be mergeable if @neheb reverts to using `TagLib::List` rather than `std::List` (including in the one added via #1122)?

**Comment by ufleisch** [comment]
I think that we should differentiate between the use of a list of `unique_ptr`s in #1122 (which I think is not worse than the code before because `unique_ptr` has no overhead - it is also not the first internal use of a `std` collection internally in the code base, so I would not interpret it as a starting signal to use `std:list` now everywhere). I would also merge the replacements of raw pointers by `unique_ptr` in this MR. But I do not want to replace implicitly shared TagLib `List`s by `std::list`s containing `shared_ptr`s - this is not equivalent at all, has more overhead and is harder to use.

**Comment by neheb** [comment]
removed std::shared_ptr commit.

File: taglib/ape/apefile.cpp [hunk 1]
```diff
@@ -52,28 +52,18 @@ namespace
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete ID3v2Header;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ape/apefile.cpp:63]: Why should these go away?

File: taglib/asf/asffile.cpp [hunk 2]
```diff
@@ -384,18 +380,13 @@ void ASF::File::FilePrivate::HeaderExtensionObject::parse(ASF::File *file, unsig
...
-    else {
-      obj = new UnknownObject(guid);
-    }
+    auto obj = [&]() mutable -> BaseObject * {
+      if(guid == metadataGuid)
+        return file->d->metadataObject = new MetadataObject();
+      if(guid == metadataLibraryGuid)
+        return file->d->metadataLibraryObject = new MetadataLibraryObject();
+      return new UnknownObject(guid);
+    }();
```
Comment (ufleisch) [inline taglib/asf/asffile.cpp:389]: You should restrict this commit to the single subject of `unique_ptr`.

File: taglib/asf/asffile.cpp [hunk 3]
```diff
@@ -617,8 +608,8 @@ void ASF::File::read()
     return;
   }
 
-  d->tag = new ASF::Tag();
-  d->properties = new ASF::Properties();
+  d->tag        = std::make_unique<ASF::Tag>();
```
Comment (ufleisch) [inline taglib/asf/asffile.cpp:611]:
```suggestion
  d->tag = std::make_unique<ASF::Tag>();
```

File: taglib/asf/asffile.cpp [hunk 4]
```diff
@@ -646,38 +637,26 @@ void ASF::File::read()
...
+        return new FilePrivate::CodecListObject();
+      if(guid == contentEncryptionGuid
+         || guid == extendedContentEncryptionGuid
+         || guid == advancedContentEncryptionGuid) {
         d->properties->setEncrypted(true);
       }
-      obj = new FilePrivate::UnknownObject(guid);
-    }
+      return new FilePrivate::UnknownObject(guid);
+    }();
```
Comment (ufleisch) [inline taglib/asf/asffile.cpp:659]: You should restrict this commit to the single subject of `unique_ptr`.

File: taglib/flac/flacfile.cpp [hunk 5]
```diff
@@ -59,14 +59,6 @@ class FLAC::File::FilePrivate
     blocks.setAutoDelete(true);
   }
 
-  ~FilePrivate()
-  {
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/flac/flacfile.cpp:68]: Why?

File: taglib/mp4/mp4file.cpp [hunk 6]
```diff
@@ -43,20 +43,9 @@ namespace
...
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete atoms;
-    delete tag;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/mp4/mp4file.cpp:55]: Why?

File: taglib/mpc/mpcfile.cpp [hunk 7]
```diff
@@ -44,28 +44,18 @@ namespace
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete ID3v2Header;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/mpc/mpcfile.cpp:55]: Why?

File: taglib/mpeg/id3v2/id3v2tag.cpp [hunk 8]
```diff
@@ -73,23 +73,14 @@ class ID3v2::Tag::TagPrivate
...
   }
 
-  ~TagPrivate()
-  {
-    delete extendedHeader;
-    delete footer;
-  }
-
-  TagPrivate(const TagPrivate &) = delete;
-  TagPrivate &operator=(const TagPrivate &) = delete;
```
Comment (ufleisch) [inline taglib/mpeg/id3v2/id3v2tag.cpp:83]: Why?

File: taglib/mpeg/mpegfile.cpp [hunk 9]
```diff
@@ -51,14 +51,6 @@ class MPEG::File::FilePrivate
   {
   }
 
-  ~FilePrivate()
-  {
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/mpeg/mpegfile.cpp:60]: Why?

File: taglib/mpeg/mpegproperties.cpp [hunk 10]
```diff
@@ -36,16 +36,7 @@ using namespace TagLib;
...
 {
 public:
-  PropertiesPrivate() = default;
-  ~PropertiesPrivate()
-  {
-    delete xingHeader;
-  }
-
-  PropertiesPrivate(const PropertiesPrivate &) = delete;
-  PropertiesPrivate &operator=(const PropertiesPrivate &) = delete;
```
Comment (ufleisch) [inline taglib/mpeg/mpegproperties.cpp:46]: Why?

File: taglib/ogg/flac/oggflacfile.cpp [hunk 11]
```diff
@@ -35,19 +35,9 @@ using TagLib::FLAC::Properties;
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete comment;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ogg/flac/oggflacfile.cpp:46]: Why?

File: taglib/ogg/oggfile.cpp [hunk 12]
```diff
@@ -53,19 +53,10 @@ class Ogg::File::FilePrivate
...
   }
 
-  ~FilePrivate()
-  {
-    delete firstPageHeader;
-    delete lastPageHeader;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ogg/oggfile.cpp:63]: Why?

File: taglib/ogg/opus/opusfile.cpp [hunk 13]
```diff
@@ -39,18 +39,8 @@ using namespace TagLib::Ogg;
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete comment;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ogg/opus/opusfile.cpp:50]: Why?

File: taglib/ogg/speex/speexfile.cpp [hunk 14]
```diff
@@ -39,18 +39,8 @@ using namespace TagLib::Ogg;
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete comment;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ogg/speex/speexfile.cpp:50]: Why?

File: taglib/ogg/speex/speexfile.cpp [hunk 15]
```diff
@@ -133,8 +123,8 @@ void Speex::File::read(bool readProperties)
 
   ByteVector commentHeaderData = packet(1);
 
-  d->comment = new Ogg::XiphComment(commentHeaderData);
+  d->comment                   = std::make_unique<Ogg::XiphComment>(commentHeaderData);
```
Comment (ufleisch) [inline taglib/ogg/speex/speexfile.cpp:126]:
```suggestion
  d->comment = std::make_unique<Ogg::XiphComment>(commentHeaderData);
```

File: taglib/ogg/vorbis/vorbisfile.cpp [hunk 16]
```diff
@@ -34,18 +34,8 @@ using namespace TagLib;
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete comment;
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/ogg/vorbis/vorbisfile.cpp:45]: Why?

File: taglib/riff/aiff/aifffile.cpp [hunk 17]
```diff
@@ -34,18 +34,8 @@ using namespace TagLib;
...
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete properties;
-    delete tag;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/riff/aiff/aifffile.cpp:45]: Why?

File: taglib/riff/aiff/aifffile.cpp [hunk 18]
```diff
@@ -155,7 +145,7 @@ void RIFF::AIFF::File::read(bool readProperties)
     const ByteVector name = chunkName(i);
     if(name == "ID3 " || name == "id3 ") {
       if(!d->tag) {
-        d->tag = new ID3v2::Tag(this, chunkOffset(i));
+        d->tag      = std::make_unique<ID3v2::Tag>(this, chunkOffset(i));
```
Comment (ufleisch) [inline taglib/riff/aiff/aifffile.cpp:148]:
```suggestion
        d->tag = std::make_unique<ID3v2::Tag>(this, chunkOffset(i));
```

File: taglib/trueaudio/trueaudiofile.cpp [hunk 19]
```diff
@@ -51,14 +51,6 @@ class TrueAudio::File::FilePrivate
   {
   }
 
-  ~FilePrivate()
-  {
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/trueaudio/trueaudiofile.cpp:60]: Why?

File: taglib/wavpack/wavpackfile.cpp [hunk 20]
```diff
@@ -47,23 +47,14 @@ namespace
...
 {
 public:
-  FilePrivate() = default;
-  ~FilePrivate()
-  {
-    delete properties;
-  }
-
-  FilePrivate(const FilePrivate &) = delete;
-  FilePrivate &operator=(const FilePrivate &) = delete;
```
Comment (ufleisch) [inline taglib/wavpack/wavpackfile.cpp:57]: Why?

**Review by ufleisch — COMMENTED** [review]
Sorry for the same question multiple times, but why should most but not all of these pimpl classes now be copyable? And you should change the commit message and MR title to reflect the real subject.

File: taglib/ape/apefile.cpp (hunk 1 above)
Comment (neheb) [reply taglib/ape/apefile.cpp:63]: The reason for them is to fix a clang-tidy warning regarding a non default constructor.

Comment (neheb) [reply taglib/ape/apefile.cpp:63]: Actually now that I think about it because of unique_ptr, it’s not copyable anyway.

**Pull request merged**